ext/openssl/ossl_pkey.c

```diff
@@ -457,6 +457,28 @@ ossl_pkey_check_public_key(const EVP_PKEY *pkey)
     ossl_raise(ePKeyError, "public key missing");
 }
 
+static VALUE
+ossl_pkey_match(VALUE self, VALUE other)
+{
+  int ret;
+  EVP_PKEY *selfPKey;
+  EVP_PKEY *otherPKey;
+
+  GetPKey(self, selfPKey);
+  GetPKey(other, otherPKey);
+
+  ret = EVP_PKEY_cmp(selfPKey, otherPKey);
+
+  if (ret == -1)
+    ossl_raise(rb_eTypeError, "cannot match different PKey types");
+  else if (ret == -2)
+    ossl_raise(rb_eNotImpError, "the pkey needs to implement param_cmp or pub_cmp");
```

Comment:
I think we should just use PKeyError instead of NotImplementedError. I think this kind of error is more of runtime error than a script error.

Reply:
K, just checked dsa/rsa/dh/ec cmp methods and the only one that 'possibly' returns a -2 is EC pub cmp
```
static int eckey_pub_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
{
    int r;
    const EC_GROUP *group = EC_KEY_get0_group(b->pkey.ec);
    const EC_POINT *pa = EC_KEY_get0_public_key(a->pkey.ec),
        *pb = EC_KEY_get0_public_key(b->pkey.ec);
    if (group == NULL || pa == NULL || pb == NULL)
        return -2;
    r = EC_POINT_cmp(group, pa, pb, NULL);
    if (r == 0)
        return 1;
    if (r == 1)
        return 0;
    return -2;
}
```
If it didn't then `NotImplementedError` would have been accurate. sigh...